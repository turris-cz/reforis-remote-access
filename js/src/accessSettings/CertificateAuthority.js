/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    Button,
    useAlert,
    useAPIDelete,
    API_STATE,
    formFieldsSize,
} from "foris";

import API_URLs from "API";

export const CA_STATUS = {
    MISSING: "missing",
    READY: "ready",
    GENERATING: "generating",
};

CertificateAuthority.propTypes = {
    authority: PropTypes.object.isRequired,
    accessEnabled: PropTypes.bool.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

export default function CertificateAuthority({
    authority,
    accessEnabled,
    onSuccess,
}) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority."));
        }
    }, [deleteResponse, onSuccess, setAlert]);

    if (authority.status !== CA_STATUS.READY) {
        return null;
    }

    return (
        <div className={formFieldsSize}>
            <h2>{_("Certificate Authority")}</h2>
            {accessEnabled ? (
                <p>
                    {_(
                        "You can't delete certificate authority while remote access is enabled. To delete it, you need to disable the access first."
                    )}
                </p>
            ) : (
                <>
                    <p>
                        {_(
                            "Your certificate authority is set up properly. Please note that if you delete it, all clients will have their access revoked."
                        )}
                    </p>
                    <div className="text-right">
                        <Button
                            onClick={() => deleteCA()}
                            loading={deleteResponse.state === API_STATE.SENDING}
                            forisFormSize
                        >
                            {_("Delete Certificate Authority")}
                        </Button>
                    </div>
                </>
            )}
        </div>
    );
}
