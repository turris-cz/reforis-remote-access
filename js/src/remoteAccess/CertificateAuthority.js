/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    Button, useAlert, useAPIDelete, API_STATE,
} from "foris";

import API_URLs from "API";

export const CA_STATUS = {
    MISSING: "missing",
    READY: "ready",
    GENERATING: "generating",
};

CertificateAuthority.propTypes = {
    authority: PropTypes.object.isRequired,
    accessEnabled: PropTypes.bool.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

export default function CertificateAuthority({ authority, accessEnabled, onSuccess }) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority"));
        }
    }, [deleteResponse, onSuccess, setAlert]);

    if (authority.status !== CA_STATUS.READY) {
        return null;
    }

    return (
        <>
            <h3>{_("Certificate Authority")}</h3>
            {accessEnabled
                ? <p>{_("You can't delete certificate authority while remote access is enabled. In order to delete it you need to disable the access first.")}</p>
                : (
                    <>
                        <p>{_("Your certificate authority is set up properly. Please note that if you delete it all clients will have their access revoked.")}</p>
                        <Button
                            onClick={() => deleteCA()}
                            loading={deleteResponse.state === API_STATE.SENDING}
                            forisFormSize
                        >
                            {_("Delete certificate authority")}
                        </Button>
                    </>
                )}
        </>
    );
}
