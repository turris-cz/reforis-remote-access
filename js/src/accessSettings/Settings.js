/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import { withErrorMessage, withSpinnerOnSending, withEither } from "foris";

import AuthorityMissing from "./AuthorityMissing";
import AuthorityGenerating from "./AuthorityGenerating";
import CertificateAuthority, { CA_STATUS } from "./CertificateAuthority";
import RemoteAccessForm from "./RemoteAccessForm";

Settings.propTypes = {
    authority: PropTypes.object.isRequired,
    settings: PropTypes.object.isRequired,
    onAuthoritySuccess: PropTypes.func.isRequired,
    onSettingsSuccess: PropTypes.func.isRequired,
};

function Settings({
    authority, settings, onAuthoritySuccess, onSettingsSuccess,
}) {
    return (
        <>
            <RemoteAccessForm
                settings={settings}
                onSuccess={onSettingsSuccess}
            />
            <CertificateAuthority
                authority={authority}
                accessEnabled={settings.enabled}
                onSuccess={onAuthoritySuccess}
            />
        </>
    );
}

const withMissing = withEither(
    (props) => props.authority.status === CA_STATUS.MISSING,
    AuthorityMissing,
);
const withGenerating = withEither(
    (props) => props.authority.status === CA_STATUS.GENERATING,
    AuthorityGenerating,
);
export default (
    withErrorMessage(
        withSpinnerOnSending(
            withMissing(
                withGenerating(Settings),
            ),
        ),
    ));
