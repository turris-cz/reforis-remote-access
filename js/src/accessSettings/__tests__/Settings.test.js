/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { WebSockets, API_STATE } from "foris";
import { render } from "foris/testUtils/customTestRender";

import Settings from "../Settings";
import { CA_STATUS } from "../CertificateAuthority";

describe("<Settings />", () => {
    const webSockets = new WebSockets();
    const onAuthoritySuccess = jest.fn();
    const onSettingsSuccess = jest.fn();

    function renderComponent(apiState, authority = {}, settings = {}) {
        return render(
            <Settings
                apiState={apiState}
                ws={webSockets}
                authority={authority}
                settings={settings}
                onAuthoritySuccess={onAuthoritySuccess}
                onSettingsSuccess={onSettingsSuccess}
            />
        );
    }

    it("should render missing component", () => {
        const { container } = renderComponent(API_STATE.SUCCESS, {
            status: CA_STATUS.MISSING,
        });
        expect(container).toMatchSnapshot();
    });

    it("should render generating component", () => {
        const { container } = renderComponent(API_STATE.SUCCESS, {
            status: CA_STATUS.GENERATING,
        });
        expect(container).toMatchSnapshot();
    });

    it("should render spinner", () => {
        const { container } = renderComponent(API_STATE.SENDING);
        expect(container).toMatchSnapshot();
    });

    it("should render error", () => {
        const { container } = renderComponent(API_STATE.ERROR);
        expect(container).toMatchSnapshot();
    });

    it("should render settings component", () => {
        const { container } = renderComponent(
            API_STATE.SUCCESS,
            { status: CA_STATUS.READY },
            { enabled: true, wan_enabled: false, port: 1234 }
        );
        expect(container).toMatchSnapshot();
    });
});
