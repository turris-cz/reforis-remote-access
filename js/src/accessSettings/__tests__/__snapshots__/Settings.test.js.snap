// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> should render error 1`] = `
<div>
  <p
    class="text-center text-danger"
  >
    An error occurred while fetching data.
  </p>
</div>
`;

exports[`<Settings /> should render generating component 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Generating Certificate Authority
    </h2>
    <p>
      Your certificate authority is now being generated. It usually takes a few minutes. Settings will appear here automatically once the authority is ready.
    </p>
    <div
      class="spinner-wrapper my-3 text-center"
    >
      <div
        class="spinner-border"
        role="status"
      >
        <span
          class="sr-only"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Settings /> should render missing component 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      No Certificate Authority
    </h2>
    <p>
      Currently, there is no certificate authority (CA) dedicated to remote access. A CA is required to generate access tokens to authenticate. To proceed, you need to generate it first.
    </p>
    <div
      class="text-end"
    >
      <button
        class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        type="button"
      >
        <span>
          Generate Certificate Authority
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Settings /> should render settings component 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Remote Access
    </h2>
    <form>
      <div
        class="form-check form-switch mb-3"
      >
        <input
          checked=""
          class="form-check-input"
          id="1"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="1"
        >
          Enable Remote Access
        </label>
      </div>
      <div
        class="form-check form-switch mb-3"
      >
        <input
          class="form-check-input"
          id="2"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="2"
        >
          Accessible via WAN
        </label>
        <div
          class="form-text"
        >
          <small>
            Devices in the WAN network will be able to connect to the configuration interface. Otherwise, only devices on LAN will be able to access it.
          </small>
        </div>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="3"
        >
          Port
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            id="3"
            type="number"
            value="1234"
          />
          <button
            aria-label="Increase value. Hint: Hold to increase faster."
            class="btn btn-outline-secondary"
            title="Increase value. Hint: Hold to increase faster."
            type="button"
          >
            <i
              class="fa"
            />
          </button>
          <button
            aria-label="Decrease value. Hint: Hold to decrease faster."
            class="btn btn-outline-secondary"
            title="Decrease value. Hint: Hold to decrease faster."
            type="button"
          >
            <i
              class="fa"
            />
          </button>
        </div>
        <div
          class="form-text"
        >
          <small>
            A port which will be opened for the remote configuration of this device.
          </small>
        </div>
      </div>
      <div
        class="text-end"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          data-testid="save-settings-button"
          type="submit"
        >
          <span>
            Save
          </span>
        </button>
      </div>
    </form>
  </div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Certificate Authority
    </h2>
    <p>
      You can't delete certificate authority while remote access is enabled. To delete it, you need to disable the access first.
    </p>
  </div>
</div>
`;

exports[`<Settings /> should render spinner 1`] = `
<div>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;
